server/src/main/java/com/cloud/vm/UserVmManagerImpl.java

```diff
@@ -3303,17 +3305,38 @@ public UserVm rebootVirtualMachine(RebootVMCmd cmd) throws InsufficientCapacityE
         return  null;
     }
 
+    /**
+     *  Encapsulates AllowUserExpungeRecoverVm so we can unit test checkExpungeVmPermission.
+     */
+    protected boolean getConfigAllowUserExpungeRecoverVm(Long entityOwnerId) {
+        return AllowUserExpungeRecoverVm.valueIn(entityOwnerId);
+    }
+
+    protected void checkExpungeVmPermission (Account callingAccount, Long entityOwnerId) {
+        logger.debug(String.format("Checking if [%s] has permission for expunging VMs.", callingAccount));
+        if (!_accountMgr.isAdmin(callingAccount.getId()) && !getConfigAllowUserExpungeRecoverVm(entityOwnerId)) {
```

Comment:
I wonder if we should check the setting of the caller (`callingAccount.getId()`, not the `entityOwnerId`) here.
`entityOwnerId` was used in the past

Reply:
Sure, it seems better. I've adjusted it.